# Data Exploration & Analysis

**Exploration:**
- Initial exploration of data attempting to investigate and define the variables

**Analysis:**
- Attempting to answer a guiding analysis question, with caveats

In [1]:
import pandas as pd

#### Load iPython-SQL module

In [5]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


#### Customize query output settings

In [3]:
%%html
<style>
    table {
        table-layout: auto; /* Table width adjusts to fit output columns */
        width: auto; /* Column width adjusts to cell contents */
    }

    table td, table th {
        white-space: nowrap; /* Prevent text wrapping within cells */
        text-align: left; /* Left-align cell contents */
    }
</style>

### Connect to database

In [4]:
%sql postgresql://postgres:12345@localhost/ecomm_cleanse

## Data Exploration

### `all_sessions` table 

#### Transaction and revenue data

In [ ]:
%%sql

SELECT visitid, totaltransactionrevenue, transactions, transactionrevenue, productrevenue
FROM all_sessions
WHERE totaltransactionrevenue IS NOT NULL
LIMIT 10;

 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('1479318391', Decimal('305.00'), '1', None, None),
 ('1487816197', Decimal('152.00'), '1', None, None),
 ('1473206048', Decimal('13.21'), '1', None, None),
 ('1474892406', Decimal('32.18'), '1', None, None),
 ('1481645890', Decimal('742.48'), '1', None, None),
 ('1495668283', Decimal('26.82'), '1', None, None),
 ('1488667799', Decimal('747.00'), '1', None, None),
 ('1494587162', Decimal('16.99'), '1', None, None),
 ('1489992840', Decimal('200.00'), '1', Decimal('200.00'), Decimal('120.00')),
 ('1490046065', Decimal('124.00'), '1', None, None)]

- lets confirm if `transactions` is always = '1' when `totaltransactionrevenue` is non-null

In [ ]:
%%sql

SELECT COUNT(*)
FROM all_sessions
WHERE totaltransactionrevenue IS NOT NULL
    AND transactions != '1';

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(0,)]

- from the above query, we can confirm that `transactions` = '1' whenever `totaltransactionrevenue` is non-null

- lets now filter for non-null `transactionrevenue` and `productrevenue` to see how they correlate

In [ ]:
%%sql

SELECT visitid, totaltransactionrevenue, transactionrevenue, productrevenue, productquantity, productprice
FROM all_sessions
WHERE transactionrevenue IS NOT NULL
    OR productrevenue IS NOT NULL;

 * postgresql://postgres:***@localhost/ecomm_cleanse
4 rows affected.


[('1489992840', Decimal('200.00'), Decimal('200.00'), Decimal('120.00'), '1', Decimal('119.00')),
 ('1491424130', Decimal('169.97'), Decimal('169.97'), Decimal('58.66'), '1', Decimal('55.99')),
 ('1478007873', Decimal('1015.48'), Decimal('1015.48'), Decimal('176.40'), '50', Decimal('3.50')),
 ('1481693637', Decimal('1005.50'), Decimal('1005.50'), Decimal('60.37'), '1', Decimal('59.99'))]

- from the above query we can make the following correlations:
    - `transactionrevenue` and `productrevenue` are only non-null when `totaltransactionrevenue` is non-null
    - `transactionrevenue` = `totaltransactionrevenue` in all occurences
    - `productrevenue` ~ `productquantity` x `productprice`
        - the small differences here could be tax, transaction fees, or some other variable not captured in this table, but represents a data quality issue

Were there any unique visitors (`fullvisitorid`) involved in more than one transaction?

In [ ]:
%%sql 

SELECT COUNT(*) AS num_transactions, fullvisitorid
FROM all_sessions
WHERE transactions IS NOT NULL
GROUP BY fullvisitorid
HAVING COUNT(*) > 1
ORDER BY num_transactions DESC;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(2, '3764227345226401562')]

the only unique visitor (`fullvisitorid`) who made more than one purchase is `3764227345226401562`

In [ ]:
%%sql

SELECT * FROM all_sessions WHERE fullvisitorid = '3764227345226401562';

 * postgresql://postgres:***@localhost/ecomm_cleanse
2 rows affected.


[('3764227345226401562', 'Direct', '0', 'United States', 'not available in demo dataset', Decimal('124.00'), '1', '342', '12', None, datetime.date(2017, 3, 20), '1490046065', 'PAGE', None, Decimal('249.00'), None, 'GGOEGAAX0794', 'Nest® Learning Thermostat 3rd Gen-USA', 'Home/Nest/Nest-USA/', '(not set)', 'USD', None, None, 'Nest-USA', '/google+redesign/', '0', '1', None),
 ('3764227345226401562', 'Direct', '342254', 'United States', 'not available in demo dataset', Decimal('124.00'), '1', '342', '12', None, datetime.date(2017, 3, 20), '1490046065', 'PAGE', None, Decimal('119.00'), None, 'GGOENEBB078899', 'Nest® Cam Indoor Security Camera - USA', 'Nest-USA', 'Single Option Only', 'USD', None, 'ORD201703201986', 'Checkout Confirmation', '/ordercompleted.html', '6', '1', None)]

inspecting the two records from this `fullvisitorid` shows that it is likely a duplicate transaction, not two distinct purchases. therefore there are only 80 transactions so far. this raises doubts about the validity of the other transactions.

we can see differenences in `productsku`, `v2productname`, `v2productcategory`, and many other columns that are null in one of the two records.

- lets quickly get a count of how many records with transactions have a `transactionid`

In [ ]:
%%sql

SELECT (
    SELECT COUNT(*)
    FROM all_sessions
    WHERE transactionid IS NOT NULL
) as num_transactionids,
(
    SELECT COUNT(*)
    FROM all_sessions
    WHERE transactions IS NOT NULL
) as num_transactions;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(9, 81)]

- from the above query, we can see there's clearly an issue with missing transactionids

lets inspect the list of records with transactions and include other visit info to see if we can catch any other obvious discrepancies

In [ ]:
%%sql

SELECT * 
FROM all_sessions
WHERE transactions IS NOT NULL
ORDER BY date, time
LIMIT 10;

 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('4934955623514361329', 'Referral', '0', 'United States', 'San Francisco', Decimal('139.42'), '1', '280', '14', None, datetime.date(2016, 8, 10), '1470870890', 'PAGE', None, Decimal('12.99'), None, 'GGOEADWQ015699', 'Android Rise 14 oz Mug', 'Home/Accessories/Drinkware/', '(not set)', 'USD', None, None, 'Drinkware', '/google+redesign/', '0', '1', None),
 ('0375962687766031488', 'Organic Search', '311843', 'United States', 'not available in demo dataset', Decimal('27.49'), '1', '331', '14', None, datetime.date(2016, 8, 11), '1470980004', 'PAGE', None, Decimal('16.99'), None, 'GGOEGAAX0105', "Google Men's 100% Cotton Short Sleeve Hero Tee Black", 'Home/Shop by Brand/Google/', '(not set)', 'USD', None, None, 'Google', '/google+redesign/', '0', '1', None),
 ('4088086075239844129', 'Organic Search', '0', 'United States', 'not available in demo dataset', Decimal('13.21'), '1', '297', '11', None, datetime.date(2016, 9, 6), '1473206048', 'PAGE', None, Decimal('19.99'), None, 'GGOEGBMB073799', 'Google Zipper-front Sports Bag', 'Home/Bags/', '(not set)', 'USD', None, None, 'Bags', '/google+redesign/', '0', '1', None),
 ('5795630000736964663', 'Referral', '72872', 'United States', 'Mountain View', Decimal('13.39'), '1', '473', '18', None, datetime.date(2016, 9, 16), '1474057380', 'PAGE', None, Decimal('13.59'), None, 'GGOEGEGQ024299', 'Grip Kit Cable Organizer', 'Home/Electronics/Electronics Accessories/', '(not set)', 'USD', None, None, 'Electronics Accessories', '/google+redesign/', '0', '1', None),
 ('803888563485194008', 'Referral', '8713', 'United States', 'not available in demo dataset', Decimal('32.18'), '1', '422', '10', None, datetime.date(2016, 9, 26), '1474892406', 'PAGE', None, Decimal('15.19'), None, 'GGOEGAAX0293', "Android Women's Short Sleeve Tri-blend Badge Tee Light Grey", "Home/Apparel/Women's/Women's-T-Shirts/", '(not set)', 'USD', None, None, "Women's-T-Shirts", '/google+redesign/', '0', '1', None),
 ('8293848749722274504', 'Direct', '10864', 'United States', 'New York', Decimal('32.99'), '1', '1512', '11', None, datetime.date(2016, 9, 28), '1475088529', 'PAGE', None, Decimal('23.99'), None, 'GGOEGAAX0610', 'Google Onesie Red/Graphite', "Home/Apparel/Kid's/Kid's-Infant/", '(not set)', 'USD', None, None, "Kid's-Infant", '/google+redesign/', '0', '1', None),
 ('7524696809408974105', 'Organic Search', '71478', 'United States', 'Houston', Decimal('38.98'), '1', '508', '18', None, datetime.date(2016, 10, 7), '1475869577', 'PAGE', None, Decimal('24.99'), None, 'GGOEGDHQ015399', '26 oz Double Wall Insulated Bottle', 'Home/Drinkware/', '(not set)', 'USD', None, None, 'Drinkware | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('1250309821000380040', 'Organic Search', '270744', 'United States', 'Columbus', Decimal('21.99'), '1', '368', '14', None, datetime.date(2016, 10, 11), '1476201122', 'PAGE', '1', Decimal('18.99'), None, 'GGOEGAAJ032617', "Google Men's Short Sleeve Badge Tee Charcoal", '(not set)', ' 2XL', 'USD', None, None, 'Checkout Your Information', '/yourinfo.html', '5', '1', 'Billing and Shipping'),
 ('7078772937328563719', 'Organic Search', '0', 'United States', 'not available in demo dataset', Decimal('38.99'), '1', '442', '15', None, datetime.date(2016, 10, 14), '1476502145', 'PAGE', None, Decimal('37.99'), None, 'GGOEGAAX0627', 'Google Infant Zip Hood Pink', "Home/Apparel/Kid's/Kid's-Infant/", '(not set)', 'USD', None, None, "Infant | Kids' Apparel | Google Merchandise Store", '/google+redesign/', '0', '1', None),
 ('9095439474545827790', 'Organic Search', '42182', 'United States', 'San Francisco', Decimal('23.99'), '1', '270', '15', None, datetime.date(2016, 10, 20), '1476976393', 'PAGE', None, Decimal('39.99'), None, 'GGOEGAAX0313', 'Google Tri-blend Hoodie Grey', "Home/Apparel/Men's/Men's-Outerwear/", '(not set)', 'USD', None, None, "Men's Outerwear | Apparel | Google Merchandise Store", '/google+redesign/', '0', '1', None)]

- no apparent patterns of obvious duplicates from inspecting the above. (similar transactions made on the same day)

- lets see if there are duplicate visitids among records with transactions

In [ ]:
%%sql

SELECT
    visitid,
    COUNT(DISTINCT fullvisitorid) AS num_fullvisitorid
FROM all_sessions
WHERE transactions IS NOT NULL
GROUP BY visitid
HAVING COUNT(fullvisitorid) > 1
ORDER BY visitid;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[('1490046065', 1)]

- its the same as the duplicate we've already found
- lets assume that the rest of the records with transactions are not duplicates (so 80 purchases in total)

#### `visitid` patterns

In [ ]:
%%sql

-- how many visitids appear in more than one row/record
SELECT
    COUNT(*)
FROM (
    SELECT
        visitid,
        COUNT(*)
    FROM all_sessions
    GROUP BY visitid
    HAVING COUNT(*) > 1
    ORDER BY visitid
);

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(553,)]

In [ ]:
%%sql

-- inspect a visitid with multiple records
SELECT visitid, COUNT(*)
FROM all_sessions
GROUP BY visitid
HAVING COUNT(*) > 1
ORDER BY COUNT(*) DESC, visitid
LIMIT 10;

 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('1471780519', 4),
 ('1474392675', 4),
 ('1470721493', 3),
 ('1471254087', 3),
 ('1471371097', 3),
 ('1471465622', 3),
 ('1471499344', 3),
 ('1471541027', 3),
 ('1471809022', 3),
 ('1472711033', 3)]

In [ ]:
%%sql
-- inspect the 2 modal visitids (4 records each)
SELECT *
FROM all_sessions
WHERE visitid IN ('1471780519', '1474392675')
ORDER BY date, time;

 * postgresql://postgres:***@localhost/ecomm_cleanse
8 rows affected.


[('3879633557713030804', 'Organic Search', '137434', 'United States', 'New York', None, None, '151', '6', None, datetime.date(2016, 8, 21), '1471780519', 'PAGE', None, Decimal('6.99'), None, 'GGOEYOLR018699', 'YouTube Leatherette Notebook Combo', 'Home/Office/', '(not set)', 'USD', None, None, 'Office', '/google+redesign/', '0', '1', None),
 ('3879633557713030804', 'Organic Search', '32178', 'United States', 'New York', None, None, '151', '6', None, datetime.date(2016, 8, 21), '1471780519', 'PAGE', None, Decimal('4.99'), None, 'GGOEYDHJ056099', '22 oz YouTube Bottle Infuser', 'Home/Drinkware/', '(not set)', 'USD', None, None, 'Drinkware', '/google+redesign/', '0', '1', None),
 ('3879633557713030804', 'Organic Search', '58188', 'United States', 'New York', None, None, '151', '6', None, datetime.date(2016, 8, 21), '1471780519', 'PAGE', None, Decimal('22.99'), None, 'GGOEGEHQ072499', 'Google 2200mAh Micro Charger', 'Home/Electronics/', '(not set)', 'USD', None, None, 'Electronics', '/google+redesign/', '0', '1', None),
 ('3879633557713030804', 'Organic Search', '58188', 'United States', 'New York', None, None, '151', '6', None, datetime.date(2016, 8, 21), '1471780519', 'PAGE', None, Decimal('16.99'), None, 'GGOEGESC014699', 'Aluminum Handy Emergency Flashlight', 'Home/Electronics/', '(not set)', 'USD', None, None, 'Electronics', '/google+redesign/', '0', '1', None),
 ('3288547201266617821', 'Direct', '109254', 'United States', 'not available in demo dataset', None, None, '134', '9', None, datetime.date(2016, 9, 20), '1474392675', 'PAGE', None, Decimal('119.99'), None, 'GGOEGAAX0585', "Google Women's Insulated Thermal Vest Navy", 'Home/Apparel/', '(not set)', 'USD', None, None, 'Apparel', '/store.html', '0', '1', None),
 ('3288547201266617821', 'Direct', '119992', 'United States', 'not available in demo dataset', None, None, '134', '9', None, datetime.date(2016, 9, 20), '1474392675', 'PAGE', None, Decimal('39.99'), None, 'GGOEGAAX0310', 'Google Heavyweight Long Sleeve Hero Tee Burgundy', 'Home/Apparel/', '(not set)', 'USD', None, None, 'Apparel', '/google+redesign/', '0', '1', None),
 ('3288547201266617821', 'Direct', '119992', 'United States', 'not available in demo dataset', None, None, '134', '9', None, datetime.date(2016, 9, 20), '1474392675', 'PAGE', None, Decimal('18.99'), None, 'GGOEGAAX0289', "Google Women's Short Sleeve Shirt Dark Grey", 'Home/Apparel/', '(not set)', 'USD', None, None, 'Apparel', '/google+redesign/', '0', '1', None),
 ('3288547201266617821', 'Direct', '57437', 'United States', 'not available in demo dataset', None, None, '134', '9', None, datetime.date(2016, 9, 20), '1474392675', 'PAGE', None, Decimal('21.99'), None, 'GGOEGHPA002910', 'Google Trucker Hat', 'Home/Apparel/', '(not set)', 'USD', None, None, 'Apparel', '/google+redesign/', '0', '1', None)]

- the visitid above looked at 4 different `productsku`
    - each product/page viewing gets its own record
    - `timeonsite` and `pageviews` values appear to apply to the `visitid` / whole session, and not to a pageviewing
        - `time` is still ambiguous, but has duplicate values for records where the `productsku` differs

- to determine which visits were duplicates, we would have to compare the `visitids` that appear multiple times, but only 1 distinct `productsku`

In [ ]:
%%sql
-- return visitids and count of associated records
WITH multiple_records AS (
    SELECT
        visitid,
        COUNT(*)
    FROM all_sessions
    GROUP BY visitid
    HAVING COUNT(*) > 1
)
-- return visitid that have multiple records but only 1 productsku, (potential duplicates)
SELECT visitid, COUNT(DISTINCT productsku) AS num_productsku
FROM all_sessions
WHERE visitid IN (SELECT visitid FROM multiple_records)
GROUP BY visitid
HAVING COUNT(DISTINCT productsku) = 1
ORDER BY visitid;

 * postgresql://postgres:***@localhost/ecomm_cleanse
5 rows affected.


[('1481233332', 1),
 ('1489797382', 1),
 ('1493605322', 1),
 ('1493909433', 1),
 ('1495005901', 1)]

In [ ]:
%%sql
CREATE TEMPORARY VIEW visitid_productsku_dupe AS (
-- return visitids and count of associated records
WITH multiple_records AS (
    SELECT
        visitid,
        COUNT(*)
    FROM all_sessions
    GROUP BY visitid
    HAVING COUNT(*) > 1
)
-- return visitid that have multiple records but only 1 productsku, (potential duplicates)
SELECT visitid, COUNT(DISTINCT productsku) AS num_productsku
FROM all_sessions
WHERE visitid IN (SELECT visitid FROM multiple_records)
GROUP BY visitid
HAVING COUNT(DISTINCT productsku) = 1
ORDER BY visitid
);

 * postgresql://postgres:***@localhost/ecomm_cleanse
Done.


[]

In [ ]:
%%sql

SELECT *
FROM all_sessions
WHERE visitid IN (SELECT visitid FROM visitid_productsku_dupe)
ORDER BY visitid, date, time;

 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('1704532013850009944', 'Direct', '129197', 'United States', 'Mountain View', None, None, '581', '13', None, datetime.date(2016, 12, 8), '1481233332', 'PAGE', None, Decimal('249.00'), None, 'GGOENEBJ079499', 'Nest® Learning Thermostat 3rd Gen-USA - Stainless Steel', 'Nest-USA', 'Single Option Only', 'USD', None, None, 'Payment Method', '/payment.html', '5', '2', 'Payment'),
 ('1704532013850009944', 'Direct', '35362', 'United States', 'Mountain View', None, None, '581', '13', None, datetime.date(2016, 12, 8), '1481233332', 'EVENT', '1', Decimal('249.00'), None, 'GGOENEBJ079499', 'Nest® Learning Thermostat 3rd Gen-USA - Stainless Steel', 'Home/Nest/Nest-USA/', '(not set)', 'USD', None, None, 'Nest-USA', '/google+redesign/', '3', '1', None),
 ('3390845580595254860', 'Affiliates', '20361', 'Indonesia', 'Jakarta', None, None, '40', '3', None, datetime.date(2017, 3, 17), '1489797382', 'PAGE', None, Decimal('8.99'), None, 'GGOEAOCH014199', 'Android Journal Book Set', 'Home/Shop by Brand/Android/', '(not set)', 'USD', None, None, 'Android | Shop by Brand | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('3390845580595254860', 'Affiliates', '39905', 'Indonesia', 'Jakarta', None, None, '40', '3', None, datetime.date(2017, 3, 17), '1489797382', 'PAGE', None, Decimal('8.99'), None, 'GGOEAOCH014199', 'Android Journal Book Set', 'Home/Shop by Brand/Android/', '(not set)', 'USD', None, None, 'Android | Shop by Brand | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('0868696744816935480', 'Referral', '40655', 'United States', 'Mountain View', None, None, '79', '7', None, datetime.date(2017, 4, 30), '1493605322', 'PAGE', None, Decimal('18.99'), None, 'GGOEGHPJ080110', 'Google 5-Panel Cap', 'Home/Apparel/', '(not set)', 'USD', None, None, 'Apparel | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('0868696744816935480', 'Referral', '78594', 'United States', 'Mountain View', None, None, '79', '7', None, datetime.date(2017, 4, 30), '1493605322', 'PAGE', None, Decimal('18.99'), None, 'GGOEGHPJ080110', 'Google 5-Panel Cap', 'Home/Apparel/', '(not set)', 'USD', None, None, 'Apparel | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('9921250175298945406', 'Organic Search', '108926', 'United States', 'not available in demo dataset', None, None, '160', '4', None, datetime.date(2017, 5, 4), '1493909433', 'PAGE', None, Decimal('20.99'), None, 'GGOEGAAX0355', "Google Men's Vintage Tank", "Home/Apparel/Men's/Men's-T-Shirts/", '(not set)', 'USD', None, None, "Men's T-Shirts | Apparel | Google Merchandise Store", '/google+redesign/', '0', '1', None),
 ('9921250175298945406', 'Organic Search', '159541', 'United States', 'not available in demo dataset', None, None, '160', '4', None, datetime.date(2017, 5, 4), '1493909433', 'PAGE', None, Decimal('20.99'), None, 'GGOEGAAX0355', "Google Men's Vintage Tank", "Home/Apparel/Men's/Men's-T-Shirts/", '(not set)', 'USD', None, None, "Men's T-Shirts | Apparel | Google Merchandise Store", '/google+redesign/', '0', '1', None),
 ('5355700591805934083', 'Referral', '13230', 'United States', 'Mountain View', None, None, '31', '3', None, datetime.date(2017, 5, 17), '1495005901', 'PAGE', None, Decimal('6.99'), None, 'GGOEGCBB074199', 'Google Car Clip Phone Holder', 'Home/Electronics/Electronics Accessories/', '(not set)', 'USD', None, None, 'Accessories | Electronics | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('5355700591805934083', 'Referral', '30527', 'United States', 'Mountain View', None, None, '31', '3', None, datetime.date(2017, 5, 17), '1495005901', 'PAGE', None, Decimal('6.99'), None, 'GGOEGCBB074199', 'Google Car Clip Phone Holder', 'Home/Shop by Brand/Google/', '(not set)', 'USD', None, None, 'Google | Shop by Brand | Google Merchandise Store', '/google+redesign/', '0', '1', None)]

**visitid = 1481233332**
- not a duplicate, the records differ by `type`, `productquantity`, `ecommerceactionstep`, `ecommerceactionoption`, `ecommerceactiontype`
    - this might represent a popup payment window, or add to cart action

**visitid IN (1489797382, 1493605322, 1493909433, 1495005901)**
- only differs by `time`, the remaining columns are identical. these might represent true duplicate records
    - if this is the case, we can remove records that only differ by `time` column, however, its uncertain which `time` value to keep

- lets investigate how `time` and `timeonsite` correlate to determine what `time` represents

In [ ]:
%%sql
-- isolate visitids with multiple records
WITH multiple_visitids AS (
    SELECT visitid, COUNT(*)
    FROM all_sessions
    GROUP BY visitid
    HAVING COUNT(*) > 1
    ORDER BY COUNT(*) DESC, visitid
)
-- return the time and timeonsite values for each visitid
SELECT visitid, time, timeonsite, pageviews--, v2productname, productsku
FROM all_sessions
WHERE visitid IN (SELECT visitid FROM multiple_visitids)
ORDER BY visitid, time
LIMIT 30;

 * postgresql://postgres:***@localhost/ecomm_cleanse
30 rows affected.


[('1470042235', '0', '41', '4'),
 ('1470042235', '40817', '41', '4'),
 ('1470093166', '29826', '91', '6'),
 ('1470093166', '75635', '91', '6'),
 ('1470152438', '166779', '186', '7'),
 ('1470152438', '4232', '186', '7'),
 ('1470157307', '0', '111', '4'),
 ('1470157307', '28506', '111', '4'),
 ('1470174905', '72451', '105', '6'),
 ('1470174905', '72451', '105', '6'),
 ('1470222606', '1298874', '1368', '7'),
 ('1470222606', '1319892', '1368', '7'),
 ('1470250539', '28349', '35', '5'),
 ('1470250539', '6773', '35', '5'),
 ('1470266781', '82140', '107', '8'),
 ('1470266781', '82140', '107', '8'),
 ('1470285826', '0', None, '1'),
 ('1470285826', '0', None, '1'),
 ('1470295887', '152612', '178', '8'),
 ('1470295887', '91493', '178', '8'),
 ('1470333871', '0', '39', '4'),
 ('1470333871', '4409', '39', '4'),
 ('1470339528', '138545', '193', '4'),
 ('1470339528', '193119', '193', '4'),
 ('1470349168', '34295', '107', '7'),
 ('1470349168', '36167', '107', '7'),
 ('1470374447', '94418', '132', '6'),
 ('1470374447', '94418', '132', '6'),
 ('1470374662', '19249', '95', '6'),
 ('1470374662', '19249', '95', '6')]

#### `time` vs `timeonsite`

- lets explore the above data to see how `time` and `timeonsite` correlate

In [ ]:
# create dataframe from above query
df = _.DataFrame()

# filter out rows with null timeonsite values
df = df[df['timeonsite'].notnull()]

# convert time and timeonsite to numeric
df['time'] = pd.to_numeric(df['time'])
df['timeonsite'] = pd.to_numeric(df['timeonsite'])

# calculate correlation coefficient of time and timeonsite
correlation = df['time'].corr(df['timeonsite'])

print(correlation)

0.9935047889587217


- `time` and `timeonsite` are correlated almost perfectly (R = .994)
    - `time` then represents the _"timeonpage"_, for that particular product page in milliseconds, `timeonesite` then represents the duration of visit
        -  `pageviews` does not correlate to this the amount of records a `visitid` has in this table, so its likely that `all_sessions` table is meant for **product focused** interactions
            - this also explains why we can't sum `time` / 1000 for a given `visitid` to equal the `timeonsite` value, as we don't have the `time` values for each page in the total `pageviews`
                - `analytics` table contains more records, likely representing every page view, a more comprehensive collection of visitor actions, not just the product focused activity

- we still can't safely remove `visitid` duplicates with differing `time` values, we'll need to look at the `analytics` table to see which is the correct value, or if these represent clicking back and forth onto the same product page.

#### records with shared `visitid`, but differing `productsku`

In [ ]:
%%sql

SELECT * FROM all_sessions WHERE visitid = '1471044901';

 * postgresql://postgres:***@localhost/ecomm_cleanse
2 rows affected.


[('2009336688383047852', 'Organic Search', '95842', 'United States', 'San Francisco', None, None, '105', '6', None, datetime.date(2016, 8, 12), '1471044901', 'PAGE', None, Decimal('16.99'), None, 'GGOEGAAX0286', "Google Women's Short Sleeve V-Neck Tee Black", "Home/Apparel/Women's/", '(not set)', 'USD', None, None, "Women's", '/store.html', '0', '1', None),
 ('2009336688383047852', 'Organic Search', '95842', 'United States', 'San Francisco', None, None, '105', '6', None, datetime.date(2016, 8, 12), '1471044901', 'PAGE', None, Decimal('16.99'), None, 'GGOEGAAX0342', "Google Women's Hero V-Neck Tee White", "Home/Apparel/Women's/", '(not set)', 'USD', None, None, "Women's", '/store.html', '0', '1', None)]

- records are identical except for `productsku` and `v2productname`, this represents looking at 2 color options for the same product
    - have identical `time` values, assume that this is the same product page, and the visitor is viewing different product variants (note that `productvariant` is not set, another data quality issue)

#### `fullvisitorid` / `visitid` relationship

- if fullvisitorid represents a unique visitor, and visitid represents a unique visit:
    - visitids should only belong to one fullvisitorid
- lets check if any visitids have multiple fullvisitorids

In [ ]:
%%sql

SELECT
    visitid,
    COUNT(DISTINCT fullvisitorid) AS num_fullvisitorid
FROM all_sessions
GROUP BY visitid
HAVING COUNT(DISTINCT fullvisitorid) > 1
ORDER BY COUNT(DISTINCT fullvisitorid) DESC, visitid;

 * postgresql://postgres:***@localhost/ecomm_cleanse
5 rows affected.


[('1472245964', 2),
 ('1481056958', 2),
 ('1486495918', 2),
 ('1489390306', 2),
 ('1493984498', 2)]

- there are 5 `visitid`s that have multiple `fullvisitorid`s
- since this is so few, we can assume that this is data integrity issue, and that `visitid`s are meant to be unique to a `fullvisitorid`.

In [ ]:
%%sql

WITH visitid_with_multiple_fullvisitorids AS (
    SELECT
        visitid,
        COUNT(DISTINCT fullvisitorid) AS num_fullvisitorid
    FROM all_sessions
    GROUP BY visitid
    HAVING COUNT(DISTINCT fullvisitorid) > 1
)

SELECT *
FROM all_sessions
WHERE visitid IN (SELECT visitid FROM visitid_with_multiple_fullvisitorids)
ORDER BY visitid, date, time;

 * postgresql://postgres:***@localhost/ecomm_cleanse
11 rows affected.


[('1564859784344874077', 'Display', '128076', 'United States', 'San Jose', None, None, '381', '12', None, datetime.date(2016, 8, 26), '1472245964', 'PAGE', None, Decimal('16.99'), None, 'GGOEGAAX0340', "Google Men's Vintage Badge Tee Green", 'Home/Shop by Brand/Google/', '(not set)', 'USD', None, None, 'Google', '/google+redesign/', '0', '1', None),
 ('3068325344054598197', 'Referral', '142913', 'United States', 'Los Angeles', None, None, '143', '9', None, datetime.date(2016, 8, 26), '1472245964', 'PAGE', None, Decimal('20.99'), None, 'GGOEGAAX0355', "Google Men's Vintage Tank", "Home/Apparel/Men's/Men's-T-Shirts/", '(not set)', 'USD', None, None, "Men's-T-Shirts", '/google+redesign/', '0', '1', None),
 ('4779983800431086057', 'Organic Search', '10883', 'United States', 'not available in demo dataset', None, None, '11', '3', None, datetime.date(2016, 12, 6), '1481056958', 'PAGE', None, Decimal('16.99'), None, 'GGOEGAAX0278', "Google Women's Short Sleeve Hero Tee Black", 'Home/Shop by Brand/Google/', '(not set)', 'USD', None, None, 'Google | Shop by Brand | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('8461696435976971245', 'Referral', '84628', 'United States', 'Mountain View', None, None, '85', '2', None, datetime.date(2016, 12, 6), '1481056958', 'PAGE', None, Decimal('79.00'), None, 'GGOENEBQ079099', 'Nest® Protect Smoke + CO White Battery Alarm-USA', 'Home/Nest/Nest-USA/', '(not set)', 'USD', None, None, 'Nest-USA', '/google+redesign/', '0', '1', None),
 ('9883320447773379187', 'Organic Search', '0', 'France', 'not available in demo dataset', None, None, None, '1', None, datetime.date(2017, 2, 7), '1486495918', 'PAGE', None, Decimal('7.99'), None, 'GGOEYDHJ019399', '24 oz YouTube Sergeant Stripe Bottle', 'Home/Shop by Brand/YouTube/', '(not set)', 'USD', None, None, 'YouTube | Shop by Brand | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('0342510849516555866', 'Organic Search', '14060', 'Panama', 'not available in demo dataset', None, None, '380', '10', None, datetime.date(2017, 2, 7), '1486495918', 'PAGE', None, Decimal('17.99'), None, 'GGOEACCQ017299', 'Android Lunch Kit', 'Home/Shop by Brand/Android/', '(not set)', 'USD', None, None, 'Android | Shop by Brand | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('9745924027985023690', 'Direct', '191393', 'Taiwan', '(not set)', None, None, '191', '6', None, datetime.date(2017, 3, 13), '1489390306', 'PAGE', None, Decimal('1.99'), None, 'GGOEYFKQ020699', 'YouTube Custom Decals', 'Home/Accessories/', '(not set)', 'USD', None, None, 'Accessories | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('3145922112411660552', 'Direct', '46200', 'Taiwan', '(not set)', None, None, '46', '3', None, datetime.date(2017, 3, 13), '1489390306', 'PAGE', None, Decimal('1.99'), None, 'GGOEGFYQ016599', 'Foam Can and Bottle Cooler', 'Home/Accessories/Housewares/', '(not set)', 'USD', None, None, 'Housewares | Accessories | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('7210036159326579347', 'Direct', '21258', 'Finland', 'not available in demo dataset', None, None, '21', '3', None, datetime.date(2017, 5, 5), '1493984498', 'PAGE', None, Decimal('29.99'), None, 'GGOEGAAX0353', 'Google Vintage Henley Grey/Black', "Home/Apparel/Men's/", '(not set)', 'USD', None, None, "Men's Apparel | Google Merchandise Store", '/google+redesign/', '0', '1', None),
 ('039119641538796794', 'Organic Search', '21523', 'United States', 'Chicago', None, None, '44', '4', None, datetime.date(2017, 5, 5), '1493984498', 'PAGE', None, Decimal('12.99'), None, 'GGOEADWQ015699', 'Android Rise 14 oz Mug', 'Home/Shop by Brand/Android/', '(not set)', 'USD', None, None, 'Android | Shop by Brand | Google Merchandise Store', '/google+redesign/', '0', '1', None),
 ('039119641538796794', 'Organic Search', '21523', 'United States', 'Chicago', None, None, '44', '4', None, datetime.date(2017, 5, 5), '1493984498', 'PAGE', None, Decimal('24.99'), Non

- from manually inspecting the above query, the only thing these `visitid`s that are assigned to multiple `fullvisitorid`s share in common is the `date`
    - because these offending `visitid`s have different `country` values, we can infer they aren't duplicates, but instead an error with id assignment
    - assumming `visitid`s are assigned in chronological order, is possible that there was a system error during `visitid` assignment (if two distinct visitors entered the site at precisely the same time, and were both assigned the same `visitid` by error. this would be worth bringing up with the associated engineers to let them know of this error)

#### `date` vs `visitid`

In [ ]:
%%sql

-- find the highest and lowest visitid values and compare the dates

SELECT
    MIN(visitid) AS min_visitid,
    MAX(visitid) AS max_visitid,
    MIN(date) AS min_date,
    MAX(date) AS max_date
FROM all_sessions;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[('1470037277', '1501657186', datetime.date(2016, 8, 1), datetime.date(2017, 8, 1))]

In [ ]:
%%sql

SELECT visitid, date
FROM all_sessions
WHERE visitid IN ('1470037277', '1501657186')

 * postgresql://postgres:***@localhost/ecomm_cleanse
2 rows affected.


[('1501657186', datetime.date(2017, 8, 1)),
 ('1470037277', datetime.date(2016, 8, 1))]

- from the above we can see that visitids are indeed assigned in chronological order, the largest visitid is associated with the latest date, and vice versa

#### Count of unique visitors and unique visitors with a transaction

In [ ]:
%%sql

SELECT (
    SELECT COUNT(DISTINCT fullvisitorid)
    FROM all_sessions
) AS num_unique_visitors,
(
    SELECT COUNT(DISTINCT fullvisitorid)
    FROM all_sessions
    WHERE totaltransactionrevenue IS NOT NULL
) AS num_unique_visitors_with_transaction;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(14223, 80)]

#### productprice - missing data

In [ ]:
%%sql

SELECT COUNT(*) FROM all_sessions WHERE productprice = '0';

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(376,)]

- there are many records where product price is 0, it should be probably be null, as the product isn't free, the data is just missing
- this hypothesis is supported by the presence of productskus where productprice = 0

In [29]:
%sql SELECT productsku, v2productname, productprice FROM all_sessions WHERE productprice = '0' LIMIT 10;

 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('9182859', 'Google Toddler Raglan Shirt Blue Heather/Navy', Decimal('0.00')),
 ('9182752', "Google Men's Short Sleeve Performance Badge Tee Charcoal", Decimal('0.00')),
 ('9182772', "Google Women's Performance Full Zip Jacket Black", Decimal('0.00')),
 ('9182553', "YouTube Men's Vintage Henley", Decimal('0.00')),
 ('9182502', "Google Women's Yoga Jacket Black", Decimal('0.00')),
 ('9180905', "Google Men's Long Sleeve Raglan Ocean Blue", Decimal('0.00')),
 ('9182739', "Google Men's Watershed Full Zip Hoodie Grey", Decimal('0.00')),
 ('9182760', "Google Women's Insulated Thermal Vest Navy", Decimal('0.00')),
 ('9182785', "Google Women's Lightweight Microfleece Jacket", Decimal('0.00')),
 ('9182575', "Android Men's  Zip Hoodie", Decimal('0.00'))]

- these records have an associated `productsku`, so its safe to assume that `productprice` = 0 is actually missing data

### `analytics` table

#### Transaction and revenue data

- the two columns associated with potential transactions appear to be `revenue` and `unitssold`
- lets filter based on these

In [ ]:
%%sql

SELECT *
FROM analytics
WHERE unitssold IS NOT NULL OR revenue IS NOT NULL
ORDER BY visitid, date, visitnumber
LIMIT 20;

 * postgresql://postgres:***@localhost/ecomm_cleanse
20 rows affected.


[('16', '1493621769', '1493622340', datetime.date(2017, 5, 1), '8178386963876992406', 'Referral', 'Not Socially Engaged', '1', '2', None, None, Decimal('157.00'), Decimal('149.00')),
 ('1', '1493621869', '1493622050', datetime.date(2017, 5, 1), '1198068166576036308', 'Organic Search', 'Not Socially Engaged', '1', '6', '180', None, None, Decimal('14.69')),
 ('2', '1493622552', '1493622552', datetime.date(2017, 5, 1), '6161000104029164017', 'Organic Search', 'Not Socially Engaged', '1', '10', '1078', None, None, Decimal('2.99')),
 ('1', '1493622674', '1493622674', datetime.date(2017, 5, 1), '3981344185539559220', 'Organic Search', 'Not Socially Engaged', '1', '13', '287', None, None, Decimal('19.99')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '4', '19', '376', None, None, Decimal('119.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '3', '19', '376', None, None, Decimal('237.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '2', '19', '376', None, None, Decimal('158.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '3', '19', '376', None, Decimal('242.00'), Decimal('79.00')),
 ('1', '1493629794', '1493629794', datetime.date(2017, 5, 1), '0167440527318407308', 'Organic Search', 'Not Socially Engaged', '1', '6', '359', None, None, Decimal('16.99')),
 ('2', '1493630490', '1493630490', datetime.date(2017, 5, 1), '7580688081499541991', 'Social', 'Not Socially Engaged', '1', '7', '328', None, None, Decimal('33.59')),
 ('3', '1493634816', '1493634816', datetime.date(2017, 5, 1), '0089335119897729017', 'Organic Search', 'Not Socially Engaged', '1', '5', '164', None, None, Decimal('10.99')),
 ('2', '1493638822', '1493638822', datetime.date(2017, 5, 1), '9888039975793005488', 'Social', 'Not Socially Engaged', '1', '20', '2081', None, None, Decimal('19.99')),
 ('1', '1493639737', '1493639737', datetime.date(2017, 5, 1), '2847957052786600334', 'Referral', 'Not Socially Engaged', '1', '33', '602', None, None, Decimal('5.99')),
 ('1', '1493639785', '1493639785', datetime.date(2017, 5, 1), '9203923090206958455', 'Organic Search', 'Not Socially Engaged', '1', '6', '295', None, None, Decimal('19.99')),
 ('1', '1493640195', '1493640195', datetime.date(2017, 5, 1), '0948655553332044353', 'Direct', 'Not Socially Engaged', '1', '8', '958', None, None, Decimal('24.99')),
 ('1', '1493641450', '1493641450', datetime.date(2017, 5, 1), '3985304281629561294', 'Referral', 'Not Socially Engaged', '5', '29', '642', None, None, Decimal('3.19')),
 ('1', '1493641450', '1493641450', datetime.date(2017, 5, 1), '3985304281629561294', 'Referral', 'Not Socially Engaged', '3', '29', '642', None, None, Decimal('1.59')),
 ('1', '1493641450', '1493641450', datetime.date(2017, 5, 1), '3985304281629561294', 'Referral', 'Not Socially Engaged', '5', '29', '642', None, None, Decimal('1.20')),
 ('1', '1493641450', '1493641450', datetime.date(2017, 5, 1), '3985304281629561294', 'Referral', 'Not Socially Engaged', '2', '29', '642', None, None, Decimal('10.39')),
 ('1', '1493641450', '1493641450', datetime.date(2017, 5, 1), '3985304281629561294', 'Referral', 'Not Socially Engaged', '3', '29', '642', None, None, Decimal('3.99'))]

- its unclear how we can determine whether `unitssold` or `revenue` can be assigned to a record
    - from the above query, we can see that `unitssold` gets assigned to all records associated with a `visitid`, but revenue is only assigned to a single record
        - `revenue` then might only occur precisely when the visitor finishes the transaction

- there are many records where `unitssold` is not null, yet `revenue` is null. its not clear if either is a reliable indicator of a visit resulting in a transaction
    - its possible that `unitssold` and `revenue` contain false positive, and false negatives. we don't have enough insight yet to rule out either

-- lets inspect a single visitid where there was a non-null `unitssold`, yet no `revenue`

In [ ]:
%%sql

SELECT *
FROM analytics
WHERE visitid = '1493629794'
ORDER BY visitid, date, visitnumber;

 * postgresql://postgres:***@localhost/ecomm_cleanse
7 rows affected.


[('1', '1493629794', '1493629794', datetime.date(2017, 5, 1), '0167440527318407308', 'Organic Search', 'Not Socially Engaged', '1', '6', '359', None, None, Decimal('16.99')),
 ('1', '1493629794', '1493629794', datetime.date(2017, 5, 1), '0167440527318407308', 'Organic Search', 'Not Socially Engaged', None, '6', '359', None, None, Decimal('16.99')),
 ('1', '1493629794', '1493629794', datetime.date(2017, 5, 1), '0167440527318407308', 'Organic Search', 'Not Socially Engaged', None, '6', '359', None, None, Decimal('18.99')),
 ('1', '1493629794', '1493629794', datetime.date(2017, 5, 1), '0167440527318407308', 'Organic Search', 'Not Socially Engaged', None, '6', '359', None, None, Decimal('19.99')),
 ('1', '1493629794', '1493629794', datetime.date(2017, 5, 1), '0167440527318407308', 'Organic Search', 'Not Socially Engaged', None, '6', '359', None, None, Decimal('20.99')),
 ('1', '1493629794', '1493629794', datetime.date(2017, 5, 1), '0167440527318407308', 'Organic Search', 'Not Socially Engaged', None, '6', '359', None, None, Decimal('24.99')),
 ('1', '1493629794', '1493629794', datetime.date(2017, 5, 1), '0167440527318407308', 'Organic Search', 'Not Socially Engaged', None, '6', '359', None, None, Decimal('29.99'))]

- now lets look at records associated with a `visitid` that had a non-null `revenue`

In [ ]:
%%sql

SELECT *
FROM analytics
WHERE visitid = '1493627652'
ORDER BY visitid, date, visitnumber;

 * postgresql://postgres:***@localhost/ecomm_cleanse
9 rows affected.


[('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '2', '19', '376', None, None, Decimal('158.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '3', '19', '376', None, Decimal('242.00'), Decimal('79.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '3', '19', '376', None, None, Decimal('237.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '4', '19', '376', None, None, Decimal('119.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', None, '19', '376', None, None, Decimal('119.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', None, '19', '376', None, None, Decimal('199.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', None, '19', '376', None, None, Decimal('237.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', None, '19', '376', None, None, Decimal('249.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', None, '19', '376', None, None, Decimal('79.00'))]

- from the above query, we can see various `unitssold` values can belong to a single `visitid`
- only one record has a non-null `revenue` value
    - this implies that `unitssold` is not an indicator for an actual transaction
        - maybe `unitssold` only indicates that an item was added to the visitors cart/basket?

- lets use non-null `revenue` values to determine the count of unique visitors who made a purchase
- first we'll count the number of records in `analytics` with non-null `revenue`

In [ ]:
%%sql

SELECT COUNT(*)
FROM analytics
WHERE revenue IS NOT NULL;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(13419,)]

- that's a lot of records. lets see what they look like (if revenue can be non-null for multiple records with the same `visitid`)

In [ ]:
%%sql

SELECT *
FROM analytics
WHERE revenue IS NOT NULL
ORDER BY fullvisitorid, visitid, visitnumber
LIMIT 20;

 * postgresql://postgres:***@localhost/ecomm_cleanse
20 rows affected.


[('2', '1495425682', '1495425682', datetime.date(2017, 5, 21), '0003961110741104601', 'Organic Search', 'Not Socially Engaged', '1', '20', '502', None, Decimal('5.99'), Decimal('2.99')),
 ('2', '1495425682', '1495425682', datetime.date(2017, 5, 21), '0003961110741104601', 'Organic Search', 'Not Socially Engaged', '1', '20', '502', None, Decimal('4.99'), Decimal('1.99')),
 ('2', '1501364305', '1501364305', datetime.date(2017, 7, 29), '0004478134742292937', 'Referral', 'Not Socially Engaged', '1', '27', '1004', None, Decimal('15.49'), Decimal('13.99')),
 ('2', '1501364305', '1501364305', datetime.date(2017, 7, 29), '0004478134742292937', 'Referral', 'Not Socially Engaged', '1', '27', '1004', None, Decimal('14.70'), Decimal('13.20')),
 ('2', '1501364305', '1501364305', datetime.date(2017, 7, 29), '0004478134742292937', 'Referral', 'Not Socially Engaged', '1', '27', '1004', None, Decimal('10.30'), Decimal('8.80')),
 ('2', '1501364305', '1501364305', datetime.date(2017, 7, 29), '0004478134742292937', 'Referral', 'Not Socially Engaged', '1', '27', '1004', None, Decimal('6.30'), Decimal('4.80')),
 ('1', '1501010265', '1501010265', datetime.date(2017, 7, 25), '0007177934947634478', 'Direct', 'Not Socially Engaged', '1', '16', '361', None, Decimal('15.59'), Decimal('13.59')),
 ('1', '1501010265', '1501010265', datetime.date(2017, 7, 25), '0007177934947634478', 'Direct', 'Not Socially Engaged', '1', '16', '361', None, Decimal('21.19'), Decimal('19.19')),
 ('1', '1493664333', '1493664333', datetime.date(2017, 5, 1), '0007933257389091624', 'Direct', 'Not Socially Engaged', '1', '17', '518', None, Decimal('37.59'), Decimal('33.59')),
 ('4', '1500336551', '1500336551', datetime.date(2017, 7, 17), '0010926318104555497', 'Referral', 'Not Socially Engaged', '1', '23', '518', None, Decimal('119.50'), Decimal('119.00')),
 ('4', '1500336551', '1500336551', datetime.date(2017, 7, 17), '0010926318104555497', 'Referral', 'Not Socially Engaged', '1', '23', '518', None, Decimal('149.50'), Decimal('149.00')),
 ('4', '1501262790', '1501262790', datetime.date(2017, 7, 28), '0010944918137684580', 'Referral', 'Not Socially Engaged', '8', '28', '327', None, Decimal('18.00'), Decimal('2.00')),
 ('4', '1501262790', '1501262790', datetime.date(2017, 7, 28), '0010944918137684580', 'Referral', 'Not Socially Engaged', '4', '28', '327', None, Decimal('21.16'), Decimal('4.79')),
 ('4', '1501262790', '1501262790', datetime.date(2017, 7, 28), '0010944918137684580', 'Referral', 'Not Socially Engaged', '4', '28', '327', None, Decimal('26.60'), Decimal('6.15')),
 ('6', '1501646806', '1501646806', datetime.date(2017, 8, 1), '00122544416887869', 'Direct', 'Not Socially Engaged', '2', '24', '996', None, Decimal('29.18'), Decimal('13.59')),
 ('6', '1501646806', '1501646806', datetime.date(2017, 8, 1), '00122544416887869', 'Direct', 'Not Socially Engaged', '1', '24', '996', None, Decimal('15.59'), Decimal('13.59')),
 ('6', '1501646806', '1501646806', datetime.date(2017, 8, 1), '00122544416887869', 'Direct', 'Not Socially Engaged', '3', '24', '996', None, Decimal('42.77'), Decimal('13.59')),
 ('3', '1496460676', '1496460676', datetime.date(2017, 6, 2), '0012561433643490595', 'Paid Search', 'Not Socially Engaged', '1', '30', '458', None, Decimal('4.30'), Decimal('2.80')),
 ('3', '1496460676', '1496460676', datetime.date(2017, 6, 2), '0012561433643490595', 'Paid Search', 'Not Socially Engaged', '1', '30', '458', None, Decimal('3.09'), Decimal('1.59')),
 ('3', '1498605303', '1498605303', datetime.date(2017, 6, 27), '0014262055593378383', 'Referral', 'Not Socially Engaged', '1', '11', '309', None, Decimal('15.51'), Decimal('9.51'))]

- `revenue` can have multiple values for the same `visitid`
    - these likely represent multiple different products being adding to the cart

- for now, lets just get a count of distinct `fullvisitorid`s who have at least one record showing non-null `revenue`

In [ ]:
%%sql

SELECT fullvisitorid, COUNT(*)
FROM analytics
WHERE revenue IS NOT NULL
GROUP BY fullvisitorid
HAVING COUNT(*) >= 1
ORDER BY COUNT(*) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('9681060687378784629', 51),
 ('5112369122544987822', 45),
 ('3052828106337222847', 34),
 ('1957458976293878100', 32),
 ('7311242886083854158', 28),
 ('4140559214164857738', 27),
 ('9026840718082010040', 26),
 ('7113011772090059658', 25),
 ('7463172420271311409', 24),
 ('7141139312225559597', 23)]

- are there records with non-null `revenue`, yet null `unitssold`?

In [ ]:
%%sql

SELECT COUNT(*)
FROM analytics
WHERE revenue IS NOT NULL
    AND unitssold IS NULL;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(0,)]

- from the above query, we can see there are no records where `revenue` is recorded without `unitssold` also having a non-null value

- lets check the inverse, can `revenue` be null when `unitssold` is non-null?

In [ ]:
%%sql

SELECT COUNT(*)
FROM analytics
WHERE revenue IS NULL
    AND unitssold IS NOT NULL;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(47435,)]

- from the above query we can see that `unitssold` does not indicate a transaction:
    - there are 47435 records where `revenue` is not null when there is an existing `unitssold` value
    - this likely indicates that `unitssold` gets associated with an entire `visitid`, and `revenue` only appears on the records where the transaction occured (payment/checkout page?)

- lets inspect relevant columns where `unitssold` is not null

In [ ]:
%%sql

SELECT date, visitid, pageviews, unitssold, unitprice, revenue
FROM analytics
WHERE unitssold IS NOT NULL
ORDER BY visitid, date
LIMIT 20;

 * postgresql://postgres:***@localhost/ecomm_cleanse
20 rows affected.


[(datetime.date(2017, 5, 1), '1493621769', '2', '1', Decimal('149.00'), Decimal('157.00')),
 (datetime.date(2017, 5, 1), '1493621869', '6', '1', Decimal('14.69'), None),
 (datetime.date(2017, 5, 1), '1493622552', '10', '1', Decimal('2.99'), None),
 (datetime.date(2017, 5, 1), '1493622674', '13', '1', Decimal('19.99'), None),
 (datetime.date(2017, 5, 1), '1493627652', '19', '3', Decimal('237.00'), None),
 (datetime.date(2017, 5, 1), '1493627652', '19', '3', Decimal('79.00'), Decimal('242.00')),
 (datetime.date(2017, 5, 1), '1493627652', '19', '4', Decimal('119.00'), None),
 (datetime.date(2017, 5, 1), '1493627652', '19', '2', Decimal('158.00'), None),
 (datetime.date(2017, 5, 1), '1493629794', '6', '1', Decimal('16.99'), None),
 (datetime.date(2017, 5, 1), '1493630490', '7', '1', Decimal('33.59'), None),
 (datetime.date(2017, 5, 1), '1493634816', '5', '1', Decimal('10.99'), None),
 (datetime.date(2017, 5, 1), '1493638822', '20', '1', Decimal('19.99'), None),
 (datetime.date(2017, 5, 1), '1493639737', '33', '1', Decimal('5.99'), None),
 (datetime.date(2017, 5, 1), '1493639785', '6', '1', Decimal('19.99'), None),
 (datetime.date(2017, 5, 1), '1493640195', '8', '1', Decimal('24.99'), None),
 (datetime.date(2017, 5, 1), '1493641450', '29', '5', Decimal('2.39'), None),
 (datetime.date(2017, 5, 1), '1493641450', '29', '3', Decimal('1.20'), None),
 (datetime.date(2017, 5, 1), '1493641450', '29', '3', Decimal('4.79'), None),
 (datetime.date(2017, 5, 1), '1493641450', '29', '3', Decimal('3.99'), None),
 (datetime.date(2017, 5, 1), '1493641450', '29', '5', Decimal('1.20'), None)]

- here we can see for a given `visitid`, there can be various `unitssold`, with or without a recorded `revenue`

- lets select a `visitid` from this list, and check if there are more records associated with null `unitssold`
    - `visitid` = '1493627652' has 4 records in this query, each with a different `unitprice`, and only one with a `revenue` value

In [ ]:
%%sql

SELECT * FROM analytics WHERE visitid = '1493627652'

 * postgresql://postgres:***@localhost/ecomm_cleanse
9 rows affected.


[('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '2', '19', '376', None, None, Decimal('158.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '3', '19', '376', None, Decimal('242.00'), Decimal('79.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '3', '19', '376', None, None, Decimal('237.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', '4', '19', '376', None, None, Decimal('119.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', None, '19', '376', None, None, Decimal('119.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', None, '19', '376', None, None, Decimal('199.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', None, '19', '376', None, None, Decimal('237.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', None, '19', '376', None, None, Decimal('249.00')),
 ('1', '1493627652', '1493627652', datetime.date(2017, 5, 1), '4871408819690446049', 'Direct', 'Not Socially Engaged', None, '19', '376', None, None, Decimal('79.00'))]

- it turns out there are 9 records in the table associated with `visitid` = '1493627652', if we include records with null `unitssold`
    - its likely that unitssold just represents products added to cart (what a terrible name for it in that case)
        - we would have to join it with all_sessions to see if this is the case


#### `date` range

In [ ]:
%%sql

SELECT
    MIN(date) AS min_date,
    MAX(date) AS max_date
FROM analytics;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(datetime.date(2017, 5, 1), datetime.date(2017, 8, 1))]

#### count of unique visitors (`fullvisitorid`)

In [ ]:
%%sql

SELECT COUNT(DISTINCT fullvisitorid)
FROM analytics

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(120018,)]

#### `visitid` vs `visitstarttime`

- `visitid` appears to be based on `visitstarttime` (unix timestamp)

In [26]:
%sql SELECT DISTINCT (visitid), visitstarttime FROM analytics LIMIT 10;

 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('1493620320', '1493622000'),
 ('1493620578', '1493623177'),
 ('1493621769', '1493622340'),
 ('1493621869', '1493622050'),
 ('1493622167', '1493622167'),
 ('1493622375', '1493622375'),
 ('1493622532', '1493622532'),
 ('1493622552', '1493622552'),
 ('1493622601', '1493622601'),
 ('1493622674', '1493622674')]

- lets see how many records have `visitid` = `visitstarttime` (compared to how many that differ)

In [27]:
%%sql

SELECT
    (SELECT COUNT(*) FROM analytics WHERE visitid = visitstarttime) as equal,
    (SELECT COUNT(*) FROM analytics WHERE visitid != visitstarttime) as unequal


 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(1725630, 13678)]

- considering that these columns have identical values the vast majority of the time, we can assume that this is the norm
    - records where there is a discrepancy could represent a potential data quality issue (with assigning visitids)

#### unitssold, unitprice, and revenue

- how do these columns correlate? is revenue = unitssold * unitprice?

In [ ]:
%%sql

SELECT visitid, revenue, unitssold, unitprice, (unitssold::INT * unitprice) AS productrevenue
FROM analytics
WHERE visitid NOT IN (
    SELECT visitid
    FROM all_sessions
    WHERE totaltransactionrevenue IS NOT NULL
        AND date BETWEEN '2017-05-01' AND '2017-08-01'
)
    AND revenue IS NOT NULL
LIMIT 10;

 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('1496336798', Decimal('106.48'), '12', Decimal('8.79'), Decimal('105.48')),
 ('1496337267', Decimal('15.13'), '1', Decimal('10.63'), Decimal('10.63')),
 ('1496337267', Decimal('15.69'), '1', Decimal('11.19'), Decimal('11.19')),
 ('1496340255', Decimal('122.00'), '1', Decimal('119.00'), Decimal('119.00')),
 ('1496340255', Decimal('479.00'), '4', Decimal('119.00'), Decimal('476.00')),
 ('1496342860', Decimal('153.50'), '1', Decimal('149.00'), Decimal('149.00')),
 ('1496342860', Decimal('83.50'), '1', Decimal('79.00'), Decimal('79.00')),
 ('1496343286', Decimal('16.32'), '1', Decimal('15.99'), Decimal('15.99')),
 ('1496345100', Decimal('11.63'), '1', Decimal('10.63'), Decimal('10.63')),
 ('1496345100', Decimal('120.00'), '1', Decimal('119.00'), Decimal('119.00'))]

- it appears that revenue is indeed calculated from unitssold * unitprice
    - discrepencies represent a data quality issue

#### bounce rate and timeonsite
- where `bounces` is not null, `timeonsite` is null. this is logical, since bounce rate represents users who immediately left the page
    - we can use this to determine where null timeonsite values may represent missing data


In [ ]:
%sql SELECT bounces, COUNT(*) FROM analytics GROUP BY bounces;

 * postgresql://postgres:***@localhost/ecomm_cleanse
2 rows affected.


[('1', 345370), (None, 1393938)]

In [ ]:
%sql SELECT * FROM analytics WHERE bounces = '1' AND timeonsite IS NOT NULL

 * postgresql://postgres:***@localhost/ecomm_cleanse
0 rows affected.


[]

In [ ]:
%sql SELECT COUNT(*) FROM analytics WHERE timeonsite IS NULL AND bounces IS NULL;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(1121,)]

- there are only 1121 rows that have null timeonsite values without being labelled as a bounce (having non-null `bounces` data)
    - this could be due to a number of reasons: an error labelling these visits as bounces,  an error recording the timeonsite value, or merely a user opening the link in another browser tab, and never viewing that tab before closing it
- we can investigate this further by looking at the associated visitids from the previous query


In [ ]:
%%sql

SELECT *
FROM analytics
WHERE timeonsite IS NULL
    AND bounces IS NULL
ORDER BY date DESC, visitid
LIMIT 10;

 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('1', '1501573106', '1501573106', datetime.date(2017, 8, 1), '7369973260731979146', 'Direct', 'Not Socially Engaged', None, '2', None, None, None, Decimal('16.99')),
 ('1', '1501573106', '1501573106', datetime.date(2017, 8, 1), '7369973260731979146', 'Direct', 'Not Socially Engaged', None, '2', None, None, None, Decimal('18.99')),
 ('1', '1501573106', '1501573106', datetime.date(2017, 8, 1), '7369973260731979146', 'Direct', 'Not Socially Engaged', None, '2', None, None, None, Decimal('29.99')),
 ('1', '1501573106', '1501573106', datetime.date(2017, 8, 1), '7369973260731979146', 'Direct', 'Not Socially Engaged', None, '2', None, None, None, Decimal('20.99')),
 ('1', '1501573106', '1501573106', datetime.date(2017, 8, 1), '7369973260731979146', 'Direct', 'Not Socially Engaged', None, '2', None, None, None, Decimal('34.99')),
 ('1', '1501573106', '1501573106', datetime.date(2017, 8, 1), '7369973260731979146', 'Direct', 'Not Socially Engaged', None, '2', None, None, None, Decimal('19.99')),
 ('2', '1501590031', '1501590031', datetime.date(2017, 8, 1), '1797402019723201242', 'Direct', 'Not Socially Engaged', None, '1', None, None, None, Decimal('55.99')),
 ('4', '1501604990', '1501604990', datetime.date(2017, 8, 1), '5021130510857499610', 'Referral', 'Not Socially Engaged', None, '2', None, None, None, Decimal('31.99')),
 ('4', '1501604990', '1501604990', datetime.date(2017, 8, 1), '5021130510857499610', 'Referral', 'Not Socially Engaged', None, '2', None, None, None, Decimal('119.99')),
 ('1', '1501608126', '1501608126', datetime.date(2017, 8, 1), '3209127324619309464', 'Referral', 'Not Socially Engaged', None, '2', None, None, None, Decimal('4.99'))]

- it would be more clearly indicative of a missing timeonsite value if we found null timeonesite for records where there is recorded revenue

In [ ]:
%%sql

SELECT *
FROM analytics
WHERE timeonsite IS NULL
    AND bounces IS NULL
    AND revenue IS NOT NULL
ORDER BY date, visitid;

 * postgresql://postgres:***@localhost/ecomm_cleanse
7 rows affected.


[('16', '1493621769', '1493622340', datetime.date(2017, 5, 1), '8178386963876992406', 'Referral', 'Not Socially Engaged', '1', '2', None, None, Decimal('157.00'), Decimal('149.00')),
 ('1', '1495608466', '1495609209', datetime.date(2017, 5, 24), '1950419220879967891', 'Referral', 'Not Socially Engaged', '1', '2', None, None, Decimal('82.00'), Decimal('79.00')),
 ('2', '1496617436', '1496617436', datetime.date(2017, 6, 4), '6115319654754427031', 'Referral', 'Not Socially Engaged', '1', '2', None, None, Decimal('123.00'), Decimal('119.00')),
 ('2', '1496617436', '1496617436', datetime.date(2017, 6, 4), '6115319654754427031', 'Referral', 'Not Socially Engaged', '1', '2', None, None, Decimal('7.99'), Decimal('3.99')),
 ('7', '1498892229', '1498892455', datetime.date(2017, 7, 1), '918950311295372029', 'Referral', 'Not Socially Engaged', '2', '2', None, None, Decimal('303.00'), Decimal('149.00')),
 ('4', '1499237694', '1499238000', datetime.date(2017, 7, 5), '8258339645709529752', 'Referral', 'Not Socially Engaged', '2', '2', None, None, Decimal('240.00'), Decimal('119.00')),
 ('4', '1499237694', '1499238000', datetime.date(2017, 7, 5), '8258339645709529752', 'Referral', 'Not Socially Engaged', '3', '2', None, None, Decimal('449.00'), Decimal('149.00'))]

- the above query shows 5 different unique visitors who have recorded revenue, indicating a transaction took place, yet they have NULL timeonsite values
    - since its not possible to complete a transaction without spending some time on the page, this is a clear indicator that there are issues with how timeonsite is recorded, require further investigation

### Comparing `all_sessions` and `analytics`

- to simplify the analysis, we will limit `all_sessions` to the date range that overlaps with `analytics

#### `fullvisitorid` and `visitid`

- are there `fullvisitorid`s and `visitid`s that are exclusive to `analytics` and `all_sessions` tables, respectively?

In [23]:
%%sql

-- views containing distinct ids from each table_total
CREATE OR REPLACE VIEW as_fullvisitorids AS (
    SELECT DISTINCT fullvisitorid
    FROM all_sessions
    WHERE date BETWEEN '2017-05-01' AND '2017-08-01'
);

CREATE OR REPLACE VIEW a_fullvisitorids AS (
    SELECT DISTINCT fullvisitorid
    FROM analytics
);

CREATE OR REPLACE VIEW as_visitids AS (
    SELECT DISTINCT visitid
    FROM all_sessions
    WHERE date BETWEEN '2017-05-01' AND '2017-08-01'
);

CREATE OR REPLACE VIEW a_visitids AS (
    SELECT DISTINCT visitid
    FROM analytics
);

-- views containing all ids between tables
CREATE OR REPLACE VIEW all_fullvisitorids AS (
    SELECT fullvisitorid
    FROM as_fullvisitorids
    UNION
    SELECT fullvisitorid
    FROM a_fullvisitorids
);

CREATE OR REPLACE VIEW all_visitids AS (
    SELECT visitid
    FROM as_visitids
    UNION
    SELECT visitid
    FROM a_visitids
);

-- views containing ids exclusive to each respective table_total
CREATE OR REPLACE VIEW a_exclusive_fvid AS (
    SELECT fullvisitorid
    FROM a_fullvisitorids
    EXCEPT
    SELECT fullvisitorid
    FROM as_fullvisitorids
);

CREATE OR REPLACE VIEW as_exclusive_fvid AS (
    SELECT fullvisitorid
    FROM as_fullvisitorids
    EXCEPT
    SELECT fullvisitorid
    FROM a_fullvisitorids
);

CREATE OR REPLACE VIEW a_exclusive_vid AS (
    SELECT visitid
    FROM a_visitids
    EXCEPT
    SELECT visitid
    FROM as_visitids
);

CREATE OR REPLACE VIEW as_exclusive_vid AS (
    SELECT visitid
    FROM as_visitids
    EXCEPT
    SELECT visitid
    FROM a_visitids
);

-- display counts from each view
SELECT -- all_sessions fullvisitorids
    (SELECT COUNT(*) FROM as_fullvisitorids) AS table_total,
    (SELECT COUNT(*) FROM as_exclusive_fvid) AS exclusive,
    (SELECT COUNT(*) FROM a_exclusive_fvid) AS missing,
    (SELECT COUNT(*) FROM all_fullvisitorids) AS combined_total,
    'all_sessions.fullvisitorid' AS column
UNION
SELECT -- analytics fullvisitorids
    (SELECT COUNT(*) FROM a_fullvisitorids) AS table_total,
    (SELECT COUNT(*) FROM a_exclusive_fvid) AS exclusive,
    (SELECT COUNT(*) FROM as_exclusive_fvid) AS missing,
    (SELECT COUNT(*) FROM all_fullvisitorids) AS combined_total,
    'analytics.fullvisitorid' AS column
UNION
SELECT -- all_sessions visitid
    (SELECT COUNT(*) FROM as_visitids) AS table_total,
    (SELECT COUNT(*) FROM as_exclusive_vid) AS exclusive,
    (SELECT COUNT(*) FROM a_exclusive_vid) AS missing,
    (SELECT COUNT(*) FROM all_visitids) AS combined_total,
    'all_sessions.visitid' AS column
UNION
SELECT -- analytics fullvisitorids
    (SELECT COUNT(*) FROM a_visitids) AS table_total,
    (SELECT COUNT(*) FROM a_exclusive_vid) AS exclusive,
    (SELECT COUNT(*) FROM as_exclusive_vid) AS missing,
    (SELECT COUNT(*) FROM all_visitids) AS combined_total,
    'analytics.visitid' AS column
    

 * postgresql://postgres:***@localhost/ecomm_cleanse
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
4 rows affected.


[(148642, 144982, 0, 148642, 'analytics.visitid'),
 (3660, 0, 144982, 148642, 'all_sessions.visitid'),
 (120018, 116388, 0, 120018, 'analytics.fullvisitorid'),
 (3630, 0, 116388, 120018, 'all_sessions.fullvisitorid')]

- if we filter all_sessions for the same date range as analytics, it appears that it does not have any unique visitors or visit records that don't also appear in analytics
    - this implies that all_sessions is reserved for recording more specific types of user interactions on the site
    - therefore analytics having unique records is not itself indicative of a data quality issue, it's likely that all_sessions data is aggregative analytics data
        - using this hypothesis, the high number of fullvisitorid and visitids exclusive to analytics might imply that these visits did not amount to anything worth recording into all_sessions (e.g., no transaction, no add to cart, etc)

#### Investigating a single visit

- next we'll join the relevant columns from `analytics` and `all_sessions`, filtering for a single `visitid`
    - we'll use one visitid with no associated transaction, and one with a transaction, to try to inferr what the purpose of each table is, and why so many `fullvisitorid`s and `visitid`s are not present inside `all_sessions`

In [ ]:
%%sql

SELECT DISTINCT a.fullvisitorid, a.visitid, a.date, a.pageviews, a.timeonsite, a.visitstarttime, s.totaltransactionrevenue, a.revenue, s.v2productname, s.productprice, a.unitprice, s.productquantity, a.unitssold
FROM all_sessions s
RIGHT JOIN analytics a USING (visitid, fullvisitorid, date, pageviews)
WHERE visitid = '1498687471'

 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('5153038635419747224', '1498687471', datetime.date(2017, 6, 28), '2', '9', '1498687471', None, None, "YouTube Men's Short Sleeve Hero Tee Charcoal", Decimal('18.99'), Decimal('1.99'), None, None),
 ('5153038635419747224', '1498687471', datetime.date(2017, 6, 28), '2', '9', '1498687471', None, None, "YouTube Men's Short Sleeve Hero Tee Charcoal", Decimal('18.99'), Decimal('4.99'), None, None),
 ('5153038635419747224', '1498687471', datetime.date(2017, 6, 28), '2', '9', '1498687471', None, None, "YouTube Men's Short Sleeve Hero Tee Charcoal", Decimal('18.99'), Decimal('6.99'), None, None),
 ('5153038635419747224', '1498687471', datetime.date(2017, 6, 28), '2', '9', '1498687471', None, None, "YouTube Men's Short Sleeve Hero Tee Charcoal", Decimal('18.99'), Decimal('10.99'), None, None),
 ('5153038635419747224', '1498687471', datetime.date(2017, 6, 28), '2', '9', '1498687471', None, None, "YouTube Men's Short Sleeve Hero Tee Charcoal", Decimal('18.99'), Decimal('14.99'), None, None),
 ('5153038635419747224', '1498687471', datetime.date(2017, 6, 28), '2', '9', '1498687471', None, None, "YouTube Men's Short Sleeve Hero Tee Charcoal", Decimal('18.99'), Decimal('16.99'), None, None),
 ('5153038635419747224', '1498687471', datetime.date(2017, 6, 28), '2', '9', '1498687471', None, None, "YouTube Men's Short Sleeve Hero Tee Charcoal", Decimal('18.99'), Decimal('18.99'), None, None),
 ('5153038635419747224', '1498687471', datetime.date(2017, 6, 28), '2', '9', '1498687471', None, None, "YouTube Men's Short Sleeve Hero Tee Charcoal", Decimal('18.99'), Decimal('19.99'), None, None),
 ('5153038635419747224', '1498687471', datetime.date(2017, 6, 28), '2', '9', '1498687471', None, None, "YouTube Men's Short Sleeve Hero Tee Charcoal", Decimal('18.99'), Decimal('20.99'), None, None),
 ('5153038635419747224', '1498687471', datetime.date(2017, 6, 28), '2', '9', '1498687471', None, None, "YouTube Men's Short Sleeve Hero Tee Charcoal", Decimal('18.99'), Decimal('29.99'), None, None)]

- the above result doesn't give us much insight
    - its interesting that `pageviews` doesn't correlate with the amount of records found in `analytics`

- lets confirm how many records are associated with this `visitid` in each table

In [ ]:
%%sql

SELECT (
    SELECT COUNT(*)
    FROM all_sessions
    WHERE visitid = '1498687471'
) AS all_sessions_records,
(
    SELECT COUNT(*)
    FROM analytics
    WHERE visitid = '1498687471'
) AS analytics_records


 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(1, 10)]

- this a 10:1 ratio for this `visitid`s records, `analytics`:`all_sessions`

- lets see what the average is for all `visitid`s

In [ ]:
%%sql
SELECT (
    WITH a_vid_count AS (
        SELECT visitid, COUNT(*)
        FROM analytics
        GROUP BY visitid
    )
    SELECT AVG(count)::NUMERIC(10,2)
    FROM a_vid_count
) AS avg_analytics_records_per_visit,
(
    WITH s_vid_count AS (
        SELECT visitid, COUNT(*)
        FROM all_sessions
        WHERE date BETWEEN '2017-05-01' AND '2017-08-01'
        GROUP BY visitid
    )
    SELECT AVG(count)::NUMERIC(10,2)
    FROM s_vid_count
) AS avg_all_sessions_records_per_visit


 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(Decimal('11.70'), Decimal('1.03'))]

- the average ratio is ~ 12:1, a typical `visitid` only appears once in `all_sessions`, and 12x in `analytics`

#### Comparing revenue / transaction data between tables

- now lets look at a visit that resulted in a transaction (where `all_sessions.transactions` is not null)

In [ ]:
%%sql

SELECT DISTINCT s.fullvisitorid, s.visitid, s.date, s.pageviews, s.timeonsite, a.visitstarttime, s.totaltransactionrevenue, a.revenue, s.v2productname, s.productprice, a.unitprice, s.productquantity, a.unitssold
FROM all_sessions s
FULL JOIN analytics a USING (visitid, fullvisitorid, date, pageviews)
WHERE visitid = '1495668283'

 * postgresql://postgres:***@localhost/ecomm_cleanse
17 rows affected.


[('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), Decimal('11.13'), "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('10.63'), None, '1'),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), Decimal('15.69'), "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('15.19'), None, '1'),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('0.00'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('3.99'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('4.79'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('10.39'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('10.63'), None, '1'),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('10.63'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('11.19'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('14.39'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('15.19'), None, '1'),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('15.19'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('15.99'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('18.99'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('19.19'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('19.99'), None, None),
 ('086102919761743448', '1495668283', datetime.date(2017, 5, 24), '12', '391', '1495668283', Decimal('26.82'), None, "Google Men's Vintage Badge Tee Sage", Decimal('10.63'), Decimal('23.19'), None, None)]

- revenue only shows up in 2 of the 17 records for this `visitid` in `analytics` table
    - but the two `revenue` values in `analytics` sum to `totaltransactionrevenue` in `all_sessions`
- `analytics.unitssold` appears non-null where revenue is non-null, and once again where it is null, but shares the same `unitprice`, indicating that this record is associated with the same product

- lets see if the `revenue`-`totaltransactionrevenue` holds for other `visitid`s by summing the `revenue` values for each `visitid`

In [ ]:
%%sql

SELECT a.visitid, totaltransactionrevenue, SUM(revenue) as sum_revenue
FROM analytics a
JOIN all_sessions s USING (visitid)
WHERE totaltransactionrevenue IS NOT NULL
    AND revenue IS NOT NULL
GROUP BY a.visitid, totaltransactionrevenue
ORDER BY visitid;

 * postgresql://postgres:***@localhost/ecomm_cleanse
22 rows affected.


[('1493668365', Decimal('81.96'), Decimal('46.98')),
 ('1494587162', Decimal('16.99'), Decimal('16.99')),
 ('1494613832', Decimal('8.98'), Decimal('8.98')),
 ('1494618681', Decimal('71.19'), Decimal('71.19')),
 ('1494701566', Decimal('123.00'), Decimal('123.00')),
 ('1494979981', Decimal('16.99'), Decimal('16.99')),
 ('1495027160', Decimal('61.97'), Decimal('40.32')),
 ('1495462087', Decimal('22.99'), Decimal('22.99')),
 ('1495668283', Decimal('26.82'), Decimal('26.82')),
 ('1495753948', Decimal('12.19'), Decimal('12.19')),
 ('1495892360', Decimal('57.99'), Decimal('57.99')),
 ('1495994469', Decimal('396.00'), Decimal('396.00')),
 ('1496601172', Decimal('391.00'), Decimal('391.00')),
 ('1496684037', Decimal('324.00'), Decimal('324.00')),
 ('1497025570', Decimal('35.99'), Decimal('35.99')),
 ('1497055585', Decimal('306.00'), Decimal('306.00')),
 ('1497459609', Decimal('23.99'), Decimal('23.99')),
 ('1498252156', Decimal('115.52'), Decimal('103.83')),
 ('1499007891', Decimal('63.97'), Decimal('63.98')),
 ('1499924733', Decimal('649.24'), Decimal('609.28')),
 ('1500066377', Decimal('358.00'), Decimal('358.00')),
 ('1501014103', Decimal('151.00'), Decimal('151.00'))]

- it appears that summed `analytics.revenue` is equal to `totaltransactionrevenue` in the majority of cases, this would indicate an issue where they don't agree
- lets isolate the `visitid`s that don't conform to this pattern

In [ ]:
%%sql
WITH revenue_discrepancy AS (
    WITH revenue_comparsion AS (
        SELECT a.visitid, totaltransactionrevenue, SUM(revenue) as sum_revenue
        FROM analytics a
        JOIN all_sessions s USING (visitid)
        WHERE totaltransactionrevenue IS NOT NULL
            AND revenue IS NOT NULL
        GROUP BY a.visitid, totaltransactionrevenue
        ORDER BY visitid
    )
    SELECT visitid
    FROM revenue_comparsion
    WHERE totaltransactionrevenue != sum_revenue
)

SELECT DISTINCT * FROM analytics WHERE visitid IN (SELECT visitid FROM revenue_discrepancy LIMIT 1);

 * postgresql://postgres:***@localhost/ecomm_cleanse
29 rows affected.


[('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', '1', '17', '732', None, Decimal('17.49'), Decimal('15.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', '1', '17', '732', None, Decimal('29.49'), Decimal('27.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', '1', '17', '732', None, None, Decimal('15.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', '1', '17', '732', None, None, Decimal('27.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('0.00')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('0.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('1.50')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('1.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('3.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('4.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('5.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('8.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('15.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('16.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('17.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('18.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('19.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('20.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('21.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('24.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('27.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic Search', 'Not Socially Engaged', None, '17', '732', None, None, Decimal('28.99')),
 ('1', '1493668365', '1493668365', datetime.date(2017, 5, 1), '1862546724568841127', 'Organic 

In [ ]:
17.49 / 15.99, 29.49 / 27.99, 15.99 * 2 + 27.99 * 2

(1.0938086303939962, 1.0535905680600215, 87.96)

- trying to make sense of why totaltransactionrevenue is almost double what the summed revenue is for visitid = 1493668365
- looking at the records where unitssold is 1, adding these up gets us closer to totaltransactionrevenue. perhaps revenue didn't get recorded correctly?
    - makes it uncertain which revenue value to trust. would assume the larger value is more accurate. totaltransactionrevenue would be less likely to get inflated for no reason?


- lets look at all_sessions for the same visitid = 1493668365

In [ ]:
%sql SELECT visitid, totaltransactionrevenue, productprice, productquantity, productrevenue FROM all_sessions WHERE visitid = '1493668365'

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[('1493668365', Decimal('81.96'), Decimal('21.99'), None, None)]

- its problematic that priceprice doesn't correlate with the unitprice found in analytics associated with the revenue
    - though there is a record with null revenue that matches this productprice (21.99)

#### Revenue disparities and missing data

- investigate:
    - `visitid`s that have `totaltransactionrevenue` in `all_sessions`, but no `revenue` in `analytics`
        - the inverse of above, `visitid`s that have `revenue` in `analytics` but no corresponding `totaltransactionrevenue` in `all_sessions`

In [ ]:
%%sql

SELECT DISTINCT ON(a.visitid) a.visitid, a.date, a.pageviews, s.totaltransactionrevenue, a.revenue,s.productprice, a.unitprice, s.productquantity, a.unitssold
FROM all_sessions s
FULL JOIN analytics a USING (visitid)
WHERE s.totaltransactionrevenue IS NULL
    AND a.revenue IS NOT NULL
LIMIT 10;


 * postgresql://postgres:***@localhost/ecomm_cleanse
10 rows affected.


[('1493621769', datetime.date(2017, 5, 1), '2', None, Decimal('157.00'), None, Decimal('149.00'), None, '1'),
 ('1493627652', datetime.date(2017, 5, 1), '19', None, Decimal('242.00'), None, Decimal('79.00'), None, '3'),
 ('1493645645', datetime.date(2017, 5, 1), '13', None, Decimal('71.68'), None, Decimal('33.59'), None, '2'),
 ('1493646617', datetime.date(2017, 5, 1), '15', None, Decimal('41.59'), None, Decimal('33.59'), None, '1'),
 ('1493647016', datetime.date(2017, 5, 1), '15', None, Decimal('41.59'), None, Decimal('33.59'), None, '1'),
 ('1493647067', datetime.date(2017, 5, 1), '11', None, Decimal('38.59'), None, Decimal('33.59'), None, '1'),
 ('1493647384', datetime.date(2017, 5, 1), '18', None, Decimal('37.59'), None, Decimal('33.59'), None, '1'),
 ('1493647545', datetime.date(2017, 5, 1), '18', None, Decimal('41.59'), None, Decimal('33.59'), None, '1'),
 ('1493647615', datetime.date(2017, 5, 1), '27', None, Decimal('14.49'), None, Decimal('12.99'), None, '1'),
 ('1493647743', datetime.date(2017, 5, 1), '23', None, Decimal('5.51'), None, Decimal('1.59'), None, '2')]

- there are clearly many instances where revenue is recorded in `analytics`, without any `totaltransactionrevenue` being recorded in `all_sessions`
- lets count how many

In [ ]:
%%sql

SELECT COUNT(*)
FROM (
    SELECT DISTINCT ON (a.visitid) a.visitid, a.date, a.pageviews, s.totaltransactionrevenue, a.revenue,s.productprice, a.unitprice, s.productquantity, a.unitssold
    FROM all_sessions s
    FULL JOIN analytics a USING (visitid)
    WHERE s.totaltransactionrevenue IS NULL
        AND a.revenue IS NOT NULL
)

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(6375,)]

- there are 6375 unique visits with with a non-null revenue value in analytics where there is no totaltransactionrevenue recorded in all_sessions
    - this implies that all_sessions is incomplete, and is not effectively aggregating data from analytics

- lets check if there are instances of recorded totaltransactionrevenue in all_sessions where there is no corresponding revenue recorded in analytics

In [ ]:
%%sql

SELECT DISTINCT ON (s.visitid) s.visitid, s.date, a.pageviews, s.totaltransactionrevenue, a.revenue,s.productprice, a.unitprice, s.productquantity, a.unitssold
FROM all_sessions s
FULL JOIN analytics a USING (visitid)
WHERE s.totaltransactionrevenue IS NOT NULL
    AND a.revenue IS NULL
    AND s.date BETWEEN '2017-05-01' AND '2017-08-02'
ORDER BY visitid;

 * postgresql://postgres:***@localhost/ecomm_cleanse
22 rows affected.


[('1493668365', datetime.date(2017, 5, 1), '17', Decimal('81.96'), None, Decimal('21.99'), Decimal('21.99'), None, None),
 ('1494587162', datetime.date(2017, 5, 12), '14', Decimal('16.99'), None, Decimal('24.99'), Decimal('29.99'), None, None),
 ('1494613832', datetime.date(2017, 5, 12), '15', Decimal('8.98'), None, Decimal('5.59'), Decimal('79.19'), None, None),
 ('1494618681', datetime.date(2017, 5, 12), '13', Decimal('71.19'), None, Decimal('55.99'), Decimal('64.99'), None, None),
 ('1494701566', datetime.date(2017, 5, 13), '12', Decimal('123.00'), None, Decimal('119.00'), Decimal('119.00'), None, '1'),
 ('1494979981', datetime.date(2017, 5, 16), '10', Decimal('16.99'), None, Decimal('15.99'), Decimal('15.99'), '1', '1'),
 ('1495027160', datetime.date(2017, 5, 17), '19', Decimal('61.97'), None, Decimal('19.99'), Decimal('0.00'), None, None),
 ('1495462087', datetime.date(2017, 5, 22), '15', Decimal('22.99'), None, Decimal('23.99'), Decimal('16.79'), None, None),
 ('1495668283', datetime.date(2017, 5, 24), '12', Decimal('26.82'), None, Decimal('10.63'), Decimal('19.19'), None, None),
 ('1495753948', datetime.date(2017, 5, 25), '18', Decimal('12.19'), None, Decimal('3.99'), Decimal('9.99'), None, None),
 ('1495892360', datetime.date(2017, 5, 27), '19', Decimal('57.99'), None, Decimal('2.99'), Decimal('15.99'), None, None),
 ('1495994469', datetime.date(2017, 5, 28), '12', Decimal('396.00'), None, Decimal('149.00'), Decimal('199.00'), None, None),
 ('1496601172', datetime.date(2017, 6, 4), '22', Decimal('391.00'), None, Decimal('149.00'), Decimal('149.00'), '1', '1'),
 ('1496684037', datetime.date(2017, 6, 5), '14', Decimal('324.00'), None, Decimal('79.00'), Decimal('238.00'), '1', None),
 ('1497025570', datetime.date(2017, 6, 9), '14', Decimal('35.99'), None, Decimal('16.99'), Decimal('33.99'), None, None),
 ('1497055585', datetime.date(2017, 6, 9), '10', Decimal('306.00'), None, Decimal('298.00'), Decimal('149.00'), None, None),
 ('1497459609', datetime.date(2017, 6, 14), '16', Decimal('23.99'), None, Decimal('34.99'), Decimal('29.99'), None, None),
 ('1498252156', datetime.date(2017, 6, 23), '15', Decimal('115.52'), None, Decimal('15.99'), Decimal('15.99'), None, '1'),
 ('1499007891', datetime.date(2017, 7, 2), '19', Decimal('63.97'), None, Decimal('4.99'), Decimal('18.99'), None, '1'),
 ('1499924733', datetime.date(2017, 7, 13), '48', Decimal('649.24'), None, Decimal('1.40'), Decimal('6.72'), None, None),
 ('1500066377', datetime.date(2017, 7, 14), '15', Decimal('358.00'), None, Decimal('119.00'), Decimal('119.00'), '1', '1'),
 ('1501014103', datetime.date(2017, 7, 25), '11', Decimal('151.00'), None, Decimal('149.00'), Decimal('199.00'), None, None)]

- there are also 23 unique visits that have no recorded revenue in `anlaytics`, yet have a `totaltransactionrevenue` recorded in `all_sessions`

- so unfortunately, we can't point to either `all_sessions` or `analytics` as a complete source of customer transactions. they are both incomplete
    - it also becomes much harder to determine which one to defer to if there is a discrepancy
        - for now, there are far fewer transactions that seem missing from analytics, so lets try to fill in this missing information using `totaltransactionrevenue`

In [ ]:
%%sql
SELECT DISTINCT ON (visitid) visitid, revenue
FROM analytics
WHERE visitid IN (
    SELECT DISTINCT ON (s.visitid) s.visitid
    FROM all_sessions s
    FULL JOIN analytics a USING (visitid)
    WHERE s.totaltransactionrevenue IS NOT NULL
        AND a.revenue IS NULL
        AND s.date BETWEEN '2017-05-01' AND '2017-08-02'
)
    AND revenue IS NOT NULL
ORDER BY visitid

 * postgresql://postgres:***@localhost/ecomm_cleanse
22 rows affected.


[('1493668365', Decimal('29.49')),
 ('1494587162', Decimal('16.99')),
 ('1494613832', Decimal('6.09')),
 ('1494618681', Decimal('71.19')),
 ('1494701566', Decimal('123.00')),
 ('1494979981', Decimal('16.99')),
 ('1495027160', Decimal('18.66')),
 ('1495462087', Decimal('22.99')),
 ('1495668283', Decimal('11.13')),
 ('1495753948', Decimal('12.19')),
 ('1495892360', Decimal('57.99')),
 ('1495994469', Decimal('396.00')),
 ('1496601172', Decimal('239.50')),
 ('1496684037', Decimal('82.50')),
 ('1497025570', Decimal('35.99')),
 ('1497055585', Decimal('306.00')),
 ('1497459609', Decimal('23.99')),
 ('1498252156', Decimal('30.88')),
 ('1499007891', Decimal('39.66')),
 ('1499924733', Decimal('8.84')),
 ('1500066377', Decimal('238.50')),
 ('1501014103', Decimal('151.00'))]

- as it turns out, the previous conclusion was not correct
    - the previous `all_sessions` query returned results from `analytics` where `visitid`s were paired with null `revenue`, however, i forgot that there are multiple entires for a given `visitid` in `analytics`
    - all 22 of the `visitid`s actually have a `revenue` value
        - in conclusion, there are no `totaltransactionrevenue` values in `all_sessions` that don't have corresponding `revenue` values in `analytics`
            - therefore `analytics` is the more complete table, and `all_sessions` is missing certain transactions (assuming that `revenue` is in fact indicative of a transaction taking place)

- we can easily confirm this by collecting the `visitid`s from `all_sessions` and using them to filter for non-null `revenue` records in `analytics`

In [ ]:
%%sql

SELECT COUNT(DISTINCT visitid)
FROM all_sessions
WHERE totaltransactionrevenue IS NOT NULL
    AND date BETWEEN '2017-05-01' AND '2017-08-01';

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(22,)]

In [ ]:
%%sql

SELECT COUNT(*)
FROM (
    SELECT DISTINCT ON (visitid) * FROM analytics
    WHERE visitid IN (SELECT visitid FROM all_sessions WHERE totaltransactionrevenue IS NOT NULL)
        AND revenue IS NOT NULL
)

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(22,)]

- the only way the count from the two most recent queries can be the same number, is if every `visitid` in `all_sessions` associated with a non-null `totaltransactionrevenue` has a corresponding non-null `revenue` value found in `analytics`

- now lets find out how many transactions are missing from `all_sessions` based on the the records found in `analytics`

In [ ]:
%%sql

SELECT COUNT(DISTINCT visitid)
FROM analytics
WHERE visitid NOT IN (
    SELECT visitid
    FROM all_sessions
    WHERE totaltransactionrevenue IS NOT NULL
        AND date BETWEEN '2017-05-01' AND '2017-08-01'
)
    AND revenue IS NOT NULL;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(6375,)]

- there are 6375 visits with recorded `revenue` (which we are assuming indicates a transaction), that have no associated `totaltransactionrevenue` in `all_sessions`

- how many of these `visitid`s are present in `all_sessions`, but have null `totaltransactionrevenue`?
- how many of these `visitid`s are missing entirely from `all_sessions`?

In [ ]:
%%sql

WITH analytics_revenue_discrepancy AS (
    SELECT DISTINCT ON (visitid) * FROM analytics
    WHERE visitid NOT IN (
        SELECT visitid
        FROM all_sessions
        WHERE totaltransactionrevenue IS NOT NULL
            AND date BETWEEN '2017-05-01' AND '2017-08-01'
    )
        AND revenue IS NOT NULL
)

SELECT s.visitid, s.totaltransactionrevenue, SUM(a.revenue) as total_analytics_revenue
FROM all_sessions s
JOIN analytics a USING (visitid)
WHERE s.visitid IN (SELECT visitid FROM analytics_revenue_discrepancy)
    AND a.revenue IS NOT NULL
GROUP BY visitid, totaltransactionrevenue;


 * postgresql://postgres:***@localhost/ecomm_cleanse
5 rows affected.


[('1494530759', None, Decimal('244.00')),
 ('1496874103', None, Decimal('153.00')),
 ('1499809854', None, Decimal('84.76')),
 ('1501191783', None, Decimal('32.99')),
 ('1501606687', None, Decimal('1002.78'))]

- the above `visitid`s are present in both `analytics` and `all_sessions`, but represent a discrepancy regarding the presence of a transaction
    - `all_sessions` has null `totaltransactionrevenue`
    - `analytics` has non-null `revenue`

- given that we've already established that `totaltransactionrevenue` = SUM(`revenue`) for a given `visitid`, we can draw 1 of two conclusions:
    1. that `all_sessions` is missing `totaltransactionrevenue` data for transactions that occured
    2. that `analytics.revenue` is not a true indicator of whether a transaction occured


#### `date` range discrepencies

In [ ]:
%%sql

SELECT
    MIN(date) AS min_date,
    MAX(date) AS max_date,
    'analytics' AS table_name
FROM analytics
UNION
SELECT
    MIN(date) AS min_date,
    MAX(date) AS max_date,
    'all_sessions' AS table_name
FROM all_sessions

 * postgresql://postgres:***@localhost/ecomm_cleanse
2 rows affected.


[(datetime.date(2016, 8, 1), datetime.date(2017, 8, 1), 'all_sessions'),
 (datetime.date(2017, 5, 1), datetime.date(2017, 8, 1), 'analytics')]

- `all_sessions` has a significantly larger `date` range than `analytics` (12 months, vs 3 months)

#### `fullvisitorid` and `visitid` counts

In [ ]:
%%sql

SELECT
    (SELECT COUNT(DISTINCT fullvisitorid) FROM analytics) AS num_unique_visitors,
    (SELECT COUNT(DISTINCT fullvisitorid) FROM analytics WHERE revenue IS NOT NULL) AS num_transacting_visitors,
    'analytics' AS table_name
UNION
SELECT
    (SELECT COUNT(DISTINCT fullvisitorid) FROM all_sessions) AS num_unique_visitors,
    (SELECT COUNT(DISTINCT fullvisitorid) FROM all_sessions WHERE totaltransactionrevenue IS NOT NULL) AS num_transacting_visitors,
    'all_sessions' AS table_name


 * postgresql://postgres:***@localhost/ecomm_cleanse


2 rows affected.


[(14223, 80, 'all_sessions'), (120018, 5799, 'analytics')]

- `analytics` has nearly 10x the number of unique visitors recorded vs `all_sessions`
- `analytics` has over 60x the number of unique transactions recorded vs `all_sessions`

## Conversion Rate

- What percentage of unique visitors make a purchase? (conversion rate)

### `all_sessions` table

In [6]:
%%sql

SELECT (
    SELECT COUNT(DISTINCT fullvisitorid)
    FROM all_sessions
) AS num_unique_visitors,
(
    SELECT COUNT(DISTINCT fullvisitorid)
    FROM all_sessions
    WHERE totaltransactionrevenue IS NOT NULL
) AS num_unique_visitors_with_transaction;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(14223, 80)]

In [7]:
conversion_rate = (80 / 14223) * 100

print(f'all_sessions conversation rate: {round(conversion_rate, 2)}%')

all_sessions conversation rate: 0.56%


If we assume `fullvisitorid` represents unique site visitors, then, according to the `all_sessions` table, **0.56% of unique visitors have made a purchase** (transaction)

#### `all_sessions` with filtered date range

In [ ]:
%%sql
-- find total number of unique visitors in all_sessions, filtering by analytics date range
SELECT COUNT(DISTINCT fullvisitorid)
FROM all_sessions
WHERE date BETWEEN '2017-05-01' AND '2017-08-01';

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(3630,)]

In [ ]:
%%sql
-- find total number of unique visitors in all_sessions who made transaction, filtering by analytics date range
SELECT COUNT(DISTINCT fullvisitorid)
FROM all_sessions
WHERE date BETWEEN '2017-05-01' AND '2017-08-01'
    AND totaltransactionrevenue IS NOT NULL;

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(22,)]

In [ ]:
all_sessions_filtered_conversion_rate = round(((22 / 3630) * 100), 2)

print(f'all_sessions: percentage of unique visitors who make a purchase: {all_sessions_filtered_conversion_rate}%')

all_sessions: percentage of unique visitors who make a purchase: 0.61%


- after ensuring the tables have the same date range, the conversion rate, as reported in all_sessions increases by 0.5%

### `analytics` table

In [15]:
%%sql

SELECT
    (SELECT COUNT(DISTINCT fullvisitorid) FROM analytics) AS num_unique_visitors,
    (SELECT COUNT(DISTINCT fullvisitorid) FROM analytics WHERE revenue IS NOT NULL) AS num_transacting_visitors

 * postgresql://postgres:***@localhost/ecomm_cleanse
1 rows affected.


[(120018, 5799)]

In [16]:
analytics_conversion_rate = round(((5799 / 120018) * 100), 2)

print(f'analytics conversion rate: {analytics_conversion_rate}%')

analytics conversion rate: 4.83%


- we can see from the above calculation, that the analytics table gives a very different conversion rate %, with far more unique visitors being associated with a transaction, assumming that a non-null `revenue` value indicates that a transaction took place

- given the findings in data exploration section, we can make an informed assumption about which table to defer to. Considering the number of missing records in `all_sessions` (missing `fullvisitorid`s, and missing recorded transactions for a given `visitid`), the conversion rate given by `analytics` is more likely to approximate the true value. However, the number of data quality issues found in this brief assessment indicates that this data is unreliable. In a real-world scenario, we would need to consult 